src/main/java/fi/aalto/cs/intellij/common/BuildInfo.java

```diff
@@ -0,0 +1,123 @@
+
+  public final Version version;
+
+  BuildInfo(@Nullable Properties props) {
+    Version ver = Version.fromProperties(props);
+    if (ver == null) {
+      ver = new Version(0,0,0);
+      LOGGER.error("Version information is invalid.  '{}' is used instead.", ver);
+    }
+    this.version = ver;
+  }
+
+  BuildInfo(@NotNull Resources resources) {
+    this(resources.getProperties(RESOURCE_NAME));
+  }
+
+  public static class Version {
+    private static final String PROPERTY_KEY = "version";
+
+    public final int major;
+    public final int minor;
+    public final int build;
+
+    /**
+     * Returns a {@link Version} object based on the value of {@value PROPERTY_KEY} in the given
+     * properties.
+     * @param props A {@link Properties} object, or {@code null} in which case {@code null} is also
+     *              returned.
+     * @return A {@link Version} object, or {@code null} if the version information could not be
+     *         read.
+     */
+    @Nullable
+    public static Version fromProperties(@Nullable Properties props) {
+      if (props == null) {
+        return null;
+      }
+      String version = props.getProperty(PROPERTY_KEY);
+      return Version.fromString(version);
+    }
+
+    /**
+     * Returns a {@link Version} object based on the value of the given string.
+     * @param version A version string of format "{major}.{minor}.{build}".
+     * @return A {@link Version} object, or {@code null} if {@code version} is invalid or null.
+     */
+    @Nullable
+    public static Version fromString(@Nullable String version) {
+      if (version == null) {
+        return null;
+      }
+
+      int major;
+      int minor;
+      int build;
+
+      StringSplitter splitter = new StringSplitter(version, '.');
+      try {
+        major = splitter.readNextInt();
+        minor = splitter.readNextInt();
+        build = splitter.readNextInt();
+      } catch (Exception ex) {
+        return null;
+      }
+
+      if (!splitter.finished()) {
+        return null;
+      }
+
+      return new Version(major, minor, build);
+    }
+
+    /**
+     * A constructor for {@link Version} class.
+     * @param major Major version number.
+     * @param minor Minor version number.
+     * @param build Number of the build.
+     */
+    public Version(int major, int minor, int build) {
+      this.major = major;
+      this.minor = minor;
+      this.build = build;
+    }
+
+    @Override
+    @NotNull
+    public String toString() {
+      return major + "." + minor + "." + build;
+    }
+
+    @Override
+    public int hashCode() {
+      return toString().hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      return obj instanceof BuildInfo.Version && toString().equals(obj.toString());
+    }
+  }
+
+  public static class SingletonUtility {
+    public static final BuildInfo INSTANCE = new BuildInfo(Resources.SingletonUtility.INSTANCE);
+
+    private SingletonUtility() { }
+  }
```

Comment:
I don't understand what does this do?

Reply:
Just a quite often-used (anti-?) pattern of lazy initialization. See, if
```
public static final BuildInfo INSTANCE = new BuildInfo(Resources.SingletonUtility.INSTANCE);
```
was in `BuildInfo` class (not in the nested class) it would be executed always when the program execution starts. Now, when it is static nested class it is not executed until some one calls it. In our case that is inside the constructor of `VersionCheckerActivity` (IMO, this is not intuitive at all but just how Java works).

An easier-to-read lazy singleton initialization would of course be:
```
class BuildInfo {
  private static BuildInfo INSTANCE;

  public static BuildInfo getInstance() {
    if (INSTANCE == null) {
      INSTANCE = new BuildInfo(...);
    }
    return INSTANCE;
  }
}
```
but that is not thread safe. Making it synchronized would help but many programmers avoid it because of "performance" (which, honestly, in our case would not be an issue, not even close). In early days of Java there were basically no chance to make non-locking thread-safe lazy initialization except this nested static class weirdness (sometimes referred as "Bill Pugh Singleton") and I guess it kinda remained as a pattern that coders (like me) keep repeating (because we are too lazy to think what would be more suitable solution until careful code reviewers catch us). :P

Perhaps the best solution is: we don't need lazy initialization at all so it can be moved out of the nested class. :D